### PR TITLE
feat: Support embed configuration

### PR DESCRIPTION
**Pull request by apoorv-mishra**
Enables admins to allow adding hosts for self hosted embeds like diagrams.net, grist etc...

### CODE REVIEW

File: server/models/Integration.ts [hunk 1]
```diff
@@ -54,17 +38,17 @@ export type IntegrationSettings =
 }))
 @Table({ tableName: "integrations", modelName: "integration" })
 @Fix
-class Integration extends IdModel {
+class Integration<T> extends IdModel {
   @IsIn([Object.values(IntegrationType)])
-  @Column
-  type: string;
+  @Column(DataType.STRING)
+  type: T;
```
Comment (apoorv-mishra) [inline server/models/Integration.ts:44]: Keeping it `T` so the type inference of settings below is dependent upon T(during usage, `T` is supposed to be passed as one of `IntegrationType.Embed`, `IntegrationType.Post` and `IntegrationType.Command`).

Comment (apoorv-mishra) [reply server/models/Integration.ts]: Reverted due to type recurring type incompatibility issues.

File: server/models/Integration.ts [hunk 2]
```diff
@@ -27,15 +55,18 @@ import Fix from "./decorators/Fix";
...
 
+  @IsIn([Object.values(IntegrationService)])
   @Column
   service: string;
 
   @Column(DataType.JSONB)
-  settings: Record<string, any>;
+  settings: IntegrationSettings;
 
+  @IsIn([["documents.update", "documents.publish"]])
```
Comment (apoorv-mishra) [inline server/models/Integration.ts:53]: Why was this added @tommoor ?

Comment (tommoor) [reply server/models/Integration.ts:53]: I was just adding validation for existing expectations, is it a requirement with this in place or still nullable?

Comment (apoorv-mishra) [reply server/models/Integration.ts:53]: What if no events are sent? Or are `events` always required?

File: shared/editor/embeds/Diagrams.tsx [hunk 3]
```diff
@@ -10,6 +11,16 @@ function Diagrams(props: Props) {
     ? `Diagrams.net (${params.get("title")})`
     : "Diagrams.net";
 
+  const integration = find(
+    props.integrations.orderedData,
```
Comment (apoorv-mishra) [inline shared/editor/embeds/Diagrams.tsx:15]: Not sure why `.orderedData` is needed here. Already passing normalized `integrations` [here](https://github.com/outline/outline/pull/3980/commits/c5a5c14089e476c661d454fa700e7507e6f46f7e#diff-18ef7e911d2802727b4354c37055ae38684c757f7de708d9010180f2ea98d248R179)

Comment (tommoor) [reply shared/editor/embeds/Diagrams.tsx:15]: Hmm, yea that doesn't seem to make any sense to me either

File: app/scenes/Document/components/DataLoader.tsx [hunk 4]
```diff
@@ -73,6 +75,19 @@ function DataLoader({ match, children }: Props) {
     fetchDocument();
   }, [ui, documents, document, shareId, documentSlug]);
 
+  React.useEffect(() => {
+    async function fetchIntegrations() {
+      try {
+        integrations.fetchPage({
```
Comment (tommoor) [inline app/scenes/Document/components/DataLoader.tsx:81]: I still think in the long run it might be better to just pull all these in on app load, otherwise there's a more likely race between this request completing and the document content showing.

Comment (apoorv-mishra) [reply app/scenes/Document/components/DataLoader.tsx:81]: What's the right place? `app/components/Editor.tsx`? Any pointers?

Comment (tommoor) [reply app/scenes/Document/components/DataLoader.tsx:81]: I'm not sure without checking but they don't _need_ to be required.

File: server/models/Integration.ts (hunk 2 above)
Comment (tommoor) [reply server/models/Integration.ts:53]: I'm not sure without checking but they don't _need_ to be required.

Comment (tommoor) [reply server/models/Integration.ts:53]: I'm not sure without checking but they don't _need_ to be required.

File: app/scenes/Document/components/DataLoader.tsx (hunk 4 above)
Comment (tommoor) [reply app/scenes/Document/components/DataLoader.tsx:81]: Somewhere super early in the app-boot that doesn't exist yet 😄 – here is okay for this PR, but lets make sure it doesn't load integrations every time, it should only load if they haven't been already.

File: shared/editor/embeds/Diagrams.tsx (hunk 3 above)
Comment (apoorv-mishra) [reply shared/editor/embeds/Diagrams.tsx:15]: Just realized that this is because of the name collision with `RootStore.integrations` [here](https://github.com/outline/outline/pull/3980/commits/c5a5c14089e476c661d454fa700e7507e6f46f7e#diff-4e4902c1514c28c9a67dea9579999973b380d7591a46c22f1574d7ebdde31389R78) :laughing:

File: shared/editor/embeds/index.tsx [hunk 5]
```diff
@@ -53,13 +53,31 @@ export type EmbedProps = {
...
+    let regexes = Component.ENABLED;
+
+    if (embedConfig) {
+      const hostname = embedConfig?.settings.hostname;
+
+      try {
+        const url = new URL(hostname);
+        regexes = [
+          new RegExp(
+            `^${url.protocol}//${url.host.replace(
```
Comment (apoorv-mishra) [inline shared/editor/embeds/index.tsx:71]: **Note**: Not considering `url.pathname`.

**Review by apoorv-mishra — COMMENTED** [review]
For now, modified the common `matcher` to _form_ the custom regex for self hosted embeds.

Maybe we can have a different pattern by supporting another attribute `urlRegex` or `host` in `EmbedDescriptor` which could then be mutated in `app/components/Editor.tsx` through `embedIntegrations` _already available_ via the component's `props`. Then, we can reference the regex(or host) by `this.urlRegex`(or `this.host`) within the common `matcher`.

File: shared/editor/rules/embeds.ts [hunk 6]
```diff
@@ -18,7 +19,10 @@ function isLinkClose(token: Token) {
   return token.type === "link_close";
 }
 
-export default function linksToEmbeds(embeds: EmbedDescriptor[]) {
+export default function linksToEmbeds(
```
Comment (tommoor) [inline shared/editor/rules/embeds.ts:22]: You know, I can't find any usage of this method – remove?

File: shared/editor/embeds/index.tsx [hunk 7]
```diff
@@ -53,13 +53,31 @@ export type EmbedProps = {
...
+              /\./g,
+              "\\."
+            )}/(?!proxy).*(title=\\\\w+)?`
```
Comment (tommoor) [inline shared/editor/embeds/index.tsx:74]: I like the idea of including this logic further down the stack as you have here, but I think this line is the giveaway that really it should remain in the `embeds/Diagrams.tsx` component – each embed will have it's own rules beyond the hostname that don't make sense to include here

Comment (apoorv-mishra) [reply shared/editor/embeds/index.tsx:74]: Yes! It was just an attempt to make it working. I just pushed the working version so I can improve it later, incorporating feedback. A lot of it is kind of _hardcoded_ right now. But thanks for this note since I had completely forgotten about it after initial revision :laughing: .Still looking out for better patterns.

File: shared/editor/rules/embeds.ts (hunk 6 above)
Comment (apoorv-mishra) [reply shared/editor/rules/embeds.ts:22]: `server/editor/renderToHtml.ts` and `shared/editor/nodes/Embed.tsx`?

Comment (tommoor) [reply shared/editor/rules/embeds.ts:22]: Well... I guess my code search was broken.

File: shared/editor/embeds/index.tsx [hunk 8]
```diff
@@ -53,13 +53,63 @@ export type EmbedProps = {
...
+
+    try {
+      const urlObj = new URL(this.host as string | URL);
+      regexes = [
+        new RegExp(
+          `^${urlObj.protocol}//${urlObj.host.replace(/\./g, "\\.")}${
+            // @ts-expect-error not aware of static
+            this.component.URL_PATH_REGEX
+          }`
+        ),
```
Comment (apoorv-mishra) [inline shared/editor/embeds/index.tsx:108]: Can be moved to `shared/utils`.

File: shared/editor/embeds/index.tsx [hunk 9]
```diff
@@ -53,13 +53,63 @@ export type EmbedProps = {
...
+  }
+
+  matcher(url: string): boolean | [] | RegExpMatchArray {
     // @ts-expect-error not aware of static
-    const regexes = Component.ENABLED;
+    let regexes = this.component.ENABLED;
+
+    try {
+      const urlObj = new URL(this.host as string | URL);
+      regexes = [
```
Comment (apoorv-mishra) [inline shared/editor/embeds/index.tsx:102]: Should self hosted and SaaS, both be supported together?

**Review by apoorv-mishra — COMMENTED** [review]
>For now, modified the common matcher to form the custom regex for self hosted embeds.

>Maybe we can have a different pattern by supporting another attribute urlRegex or host in EmbedDescriptor which could then be mutated in app/components/Editor.tsx through embedIntegrations already available via the component's props. Then, we can reference the regex(or host) by this.urlRegex(or this.host) within the common matcher.

_Take two_ :point_up: 

cc: @tommoor

File: shared/editor/embeds/index.tsx (hunk 7 above)
Comment (apoorv-mishra) [reply shared/editor/embeds/index.tsx]: Fixed!

File: shared/editor/embeds/index.tsx (hunk 8 above)
Comment (apoorv-mishra) [reply shared/editor/embeds/index.tsx]: Done!

File: app/components/Editor.tsx [hunk 10]
```diff
@@ -280,7 +280,16 @@ function Editor(props: Props, ref: React.RefObject<SharedEditor> | null) {
...
           onShowToast={showToast}
-          embeds={embeds}
+          embeds={embeds.map((e) => {
+            const em = find(
+              props.embedIntegrations,
+              (i) => i.service === e.component.name.toLowerCase()
+            );
+            return new EmbedDescriptor({
+              ...e,
+              host: em?.settings.hostname,
```
Comment (tommoor) [inline app/components/Editor.tsx:290]: I'd suggest we spread all settings here, `hostname` won't be the only one needed in the long run. Also should choose – `host` or `hostname`, host is better but technically doesn't include the protocol so probably just use `url` is the closest to accurate?

File: app/components/Editor.tsx [hunk 11]
```diff
@@ -280,7 +280,16 @@ function Editor(props: Props, ref: React.RefObject<SharedEditor> | null) {
           ref={mergeRefs([ref, handleRefChanged])}
           uploadFile={onUploadFile}
           onShowToast={showToast}
-          embeds={embeds}
+          embeds={embeds.map((e) => {
```
Comment (tommoor) [inline app/components/Editor.tsx:283]: Lets memoize

File: app/hooks/useAuthorizedSettingsConfig.ts [hunk 12]
```diff
@@ -170,6 +172,14 @@ const useAuthorizedSettingsConfig = () => {
         group: t("Integrations"),
         icon: WebhooksIcon,
       },
+      Diagrams: {
```
Comment (tommoor) [inline app/hooks/useAuthorizedSettingsConfig.ts:175]: I was half way through deciding when handed over to you, lets just rename all these to `Drawio` instead of `Diagrams`? I

File: shared/utils/escape.ts [hunk 13]
```diff
@@ -0,0 +1,4 @@
+// source: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions
+export const escapeRegExp = (text: string) => {
```
Comment (tommoor) [inline shared/utils/escape.ts:2]: This utility exists in lodash, please use that one

File: shared/utils/urls.ts [hunk 14]
```diff
@@ -92,3 +93,13 @@ export function sanitizeUrl(url: string | null | undefined) {
   }
   return url;
 }
+
+export function urlRegex(url: string | null | undefined): RegExp | undefined {
+  if (!url || !isUrl(url)) {
```
Comment (tommoor) [inline shared/utils/urls.ts:98]: Lets use `sanitizeUrl` inside this method to make sure we're filtering out malicious urls from being embedded in an iframe

File: app/components/Editor.tsx (hunk 10 above)
Comment (apoorv-mishra) [reply app/components/Editor.tsx:290]: Agree. `url` makes more sense. Will update that.

File: app/hooks/useAuthorizedSettingsConfig.ts (hunk 12 above)
Comment (apoorv-mishra) [reply app/hooks/useAuthorizedSettingsConfig.ts:175]: Sure!

File: shared/utils/escape.ts (hunk 13 above)
Comment (apoorv-mishra) [reply shared/utils/escape.ts:2]: Thanks for the tip :+1:

File: shared/utils/urls.ts (hunk 14 above)
Comment (apoorv-mishra) [reply shared/utils/urls.ts:98]: Yes! Just realized this. Will update.

File: app/components/Editor.tsx (hunk 11 above)
Comment (apoorv-mishra) [reply app/components/Editor.tsx:283]: Didn't understand! You mean https://lodash.com/docs/4.17.15#memoize?

Comment (tommoor) [reply app/components/Editor.tsx:283]: As this one has been converted to functional already can use `useMemo` hook – I want to make sure it isn't mapping to new objects very render

File: shared/utils/urls.ts (hunk 14 above)
Comment (apoorv-mishra) [reply shared/utils/urls.ts:98]: Fixed!

File: shared/utils/escape.ts (hunk 13 above)
Comment (apoorv-mishra) [reply shared/utils/escape.ts]: Fixed!

File: app/components/Editor.tsx (hunk 11 above)
Comment (apoorv-mishra) [reply app/components/Editor.tsx]: Fixed!

File: app/components/Editor.tsx (hunk 10 above)
Comment (apoorv-mishra) [reply app/components/Editor.tsx]: Fixed!

File: app/hooks/useAuthorizedSettingsConfig.ts (hunk 12 above)
Comment (apoorv-mishra) [reply app/hooks/useAuthorizedSettingsConfig.ts]: Fixed!

File: shared/editor/embeds/index.tsx (hunk 9 above)
Comment (apoorv-mishra) [reply shared/editor/embeds/index.tsx]: Made modifications to support both.

File: app/scenes/Settings/Drawio.tsx [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+      url: integration?.settings.url,
+    },
+  });
+
+  const handleSubmit = React.useCallback(
+    async (data: FormData) => {
+      if (integration) {
+        await integration.save({
+          settings: {
+            url: data.url,
```
Comment (apoorv-mishra) [inline app/scenes/Settings/Drawio.tsx:49]: There should be a stricter validation for `url` here.

Comment (apoorv-mishra) [reply app/scenes/Settings/Drawio.tsx:49]: Added on backend.

File: app/components/Editor.tsx [hunk 16]
```diff
@@ -54,10 +54,24 @@ export type Props = Optional<
 };
 
 function Editor(props: Props, ref: React.RefObject<SharedEditor> | null) {
-  const { id, shareId, onChange, onHeadingsChange } = props;
+  const { id, shareId, onChange, onHeadingsChange, embedIntegrations } = props;
   const { documents } = useStores();
   const { showToast } = useToasts();
   const dictionary = useDictionary();
+  const embedsWithSetings = React.useMemo(
```
Comment (tommoor) [inline app/components/Editor.tsx:61]: I'd consider pulling this into a local hook, `useEmbedIntegrations` or similar, we can pull the info from the store in the hook and totally avoid passing `embedIntegrations` into this component.

File: app/scenes/Settings/Drawio.tsx [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+  url: string;
+};
+
+function Drawio() {
+  const { integrations } = useStores();
+  const { t } = useTranslation();
+
+  React.useEffect(() => {
+    integrations.fetchPage({
+      limit: 100,
```
Comment (tommoor) [inline app/scenes/Settings/Drawio.tsx:26]:
```suggestion
      limit: 100,
      type: IntegrationType.Embed,
```

File: app/scenes/Settings/Drawio.tsx [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+        });
+      } else {
+        await integrations.create({
+          type: IntegrationType.Embed,
+          service: "diagrams",
+          settings: {
+            url: data.url,
+          },
+        });
+      }
```
Comment (tommoor) [inline app/scenes/Settings/Drawio.tsx:60]:
```suggestion
      await integrations.save({
        type: IntegrationType.Embed,
        service: "diagrams",
        settings: {
          url: data.url,
        },
      });
```

I know it was my code previously, but this should work. update/create switch is taken care of in BaseStore.

File: app/scenes/Settings/Drawio.tsx [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+      title={t("Draw.io")}
+      icon={<BuildingBlocksIcon color="currentColor" />}
+    >
+      <Heading>{t("Draw.io")}</Heading>
+
+      <Text type="secondary">
+        <Trans>
+          If your team is self-hosting draw.io then adding your hosted
+          deployment url here will enable automatic embedding within documents.
+          If you are using Diagrams.net then no configuration is required.
```
Comment (tommoor) [inline app/scenes/Settings/Drawio.tsx:76]:
```suggestion
          Add your self-hosted draw.io installation url here to enable automatic 
          embedding of diagrams within documents.
```

File: app/scenes/Settings/Drawio.tsx [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+        </Trans>
+        <form onSubmit={formHandleSubmit(handleSubmit)}>
+          <p>
+            <Input
+              label={t("Draw.io Deployment")}
```
Comment (tommoor) [inline app/scenes/Settings/Drawio.tsx:81]:
```suggestion
              label={t("Draw.io deployment")}
```

File: app/scenes/Settings/Drawio.tsx [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+          },
+        });
+      }
+    },
+    [integrations, integration]
+  );
+
+  return (
+    <Scene
+      title={t("Draw.io")}
```
Comment (tommoor) [inline app/scenes/Settings/Drawio.tsx:67]:
```suggestion
      title="Draw.io"
```

Product name never needs translating

File: app/scenes/Settings/Drawio.tsx [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+      icon={<BuildingBlocksIcon color="currentColor" />}
+    >
+      <Heading>{t("Draw.io")}</Heading>
```
Comment (tommoor) [inline app/scenes/Settings/Drawio.tsx:70]:
```suggestion
      <Heading>Draw.io</Heading>
```

Product name never needs translating

File: app/scenes/Settings/Drawio.tsx [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+          If your team is self-hosting draw.io then adding your hosted
+          deployment url here will enable automatic embedding within documents.
+          If you are using Diagrams.net then no configuration is required.
+        </Trans>
+        <form onSubmit={formHandleSubmit(handleSubmit)}>
+          <p>
+            <Input
+              label={t("Draw.io Deployment")}
+              placeholder={"https://app.diagrams.net/"}
+              pattern="https://.*"
```
Comment (tommoor) [inline app/scenes/Settings/Drawio.tsx:83]:
```suggestion
              pattern="https?://.*"
```

Probably a good idea to support non-secure deployments

File: app/scenes/Settings/Drawio.tsx [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+      } else {
+        await integrations.create({
+          type: IntegrationType.Embed,
+          service: "diagrams",
+          settings: {
+            url: data.url,
+          },
+        });
+      }
+    },
```
Comment (tommoor) [inline app/scenes/Settings/Drawio.tsx:61]: Need to show a toast message here too otherwise it's unclear that settings saved, something like:

```
    showToast(t("Settings saved"), {
      type: "success",
    });
```

File: server/routes/api/integrations.ts [hunk 25]
```diff
@@ -43,20 +80,30 @@ router.post("integrations.update", auth(), async (ctx) => {
...
+    integration.set({
+      events: events.filter((event: string) =>
+        ["documents.update", "documents.publish"].includes(event)
+      ),
+    });
+  }
+
+  integration.set({
+    settings,
+  });
```
Comment (tommoor) [inline server/routes/api/integrations.ts:97]:
```suggestion
  integration.settings = settings;
```

Haven't used this `set` api before, seems unneccessary? :P

File: shared/editor/embeds/Diagrams.tsx [hunk 26]
```diff
@@ -3,43 +3,36 @@ import Frame from "../components/Frame";
 import Image from "../components/Image";
 import { EmbedProps as Props } from ".";
 
-const URL_REGEX = /^https:\/\/viewer\.diagrams\.net\/(?!proxy).*(title=\\w+)?/;
+function Diagrams(props: Props) {
+  const embedUrl = props.attrs.matches[0];
+  const params = new URL(embedUrl).searchParams;
+  const title = params.get("title")
```
Comment (tommoor) [inline shared/editor/embeds/Diagrams.tsx:9]: As it displays below the embed can we show "Draw.io" as title when the embed `settings` is present?

**Review by tommoor — COMMENTED** [review]
I think that's everything, should be good to merge once these small things are resolved.

File: app/components/Editor.tsx (hunk 16 above)
Comment (apoorv-mishra) [reply app/components/Editor.tsx:61]: > we can pull the info from the store in the hook

_Pull from the server from within the hook_? As it's currently done in [`DataLoader.tsx`](https://github.com/outline/outline/pull/3980/files#diff-18ef7e911d2802727b4354c37055ae38684c757f7de708d9010180f2ea98d248R82)? Or keep the API call in `DataLoader.tsx` and just do `return integrations.orderedData(...)` in the hook? The latter should work too since `DataLoader.tsx` mounts before `Editor.tsx`.

Comment (tommoor) [reply app/components/Editor.tsx:61]: I was thinking about that, we could co-locate the data loading yes. Later we may want to support these integrations on public share links, that _may_ mean moving the data loading again. But it would be neat to include it in the hook for now

File: app/scenes/Settings/Drawio.tsx (hunk 17 above)
Comment (apoorv-mishra) [reply app/scenes/Settings/Drawio.tsx:26]: :point_down: seems better?

```suggestion
    service: "diagrams",
```

File: server/routes/api/integrations.ts (hunk 25 above)
Comment (apoorv-mishra) [reply server/routes/api/integrations.ts:97]: Yes! Makes more sense when multiple attributes need to be set.

File: shared/editor/embeds/Diagrams.tsx (hunk 26 above)
Comment (apoorv-mishra) [reply shared/editor/embeds/Diagrams.tsx:9]: Sure!

File: app/scenes/Settings/Drawio.tsx (hunk 17 above)
Comment (apoorv-mishra) [reply app/scenes/Settings/Drawio.tsx]: Fixed!

File: app/scenes/Settings/Drawio.tsx (hunk 18 above)
Comment (apoorv-mishra) [reply app/scenes/Settings/Drawio.tsx]: Fixed!
(Needed `id: integration?.id` as well for `update`)

File: app/scenes/Settings/Drawio.tsx (hunk 19 above)
Comment (apoorv-mishra) [reply app/scenes/Settings/Drawio.tsx]: Fixed!

File: app/scenes/Settings/Drawio.tsx (hunk 20 above)
Comment (apoorv-mishra) [reply app/scenes/Settings/Drawio.tsx]: Fixed!

File: app/scenes/Settings/Drawio.tsx (hunk 21 above)
Comment (apoorv-mishra) [reply app/scenes/Settings/Drawio.tsx]: Fixed!

File: app/scenes/Settings/Drawio.tsx (hunk 22 above)
Comment (apoorv-mishra) [reply app/scenes/Settings/Drawio.tsx]: Fixed!

File: app/components/Editor.tsx (hunk 16 above)
Comment (apoorv-mishra) [reply app/components/Editor.tsx]: Fixed!

File: app/scenes/Settings/Drawio.tsx (hunk 23 above)
Comment (apoorv-mishra) [reply app/scenes/Settings/Drawio.tsx]: Fixed!

File: app/scenes/Settings/Drawio.tsx (hunk 24 above)
Comment (apoorv-mishra) [reply app/scenes/Settings/Drawio.tsx:68]: Fixed!
(Added for `type: "error"` as well)

File: server/routes/api/integrations.ts (hunk 25 above)
Comment (apoorv-mishra) [reply server/routes/api/integrations.ts]: Fixed!

File: shared/editor/embeds/Diagrams.tsx (hunk 26 above)
Comment (apoorv-mishra) [reply shared/editor/embeds/Diagrams.tsx:11]: Fixed!

File: app/hooks/useEmbedIntegrations.ts [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+  >();
+
+  React.useEffect(() => {
+    async function fetchEmbedIntegrations() {
+      try {
+        await integrations.fetchPage({
+          limit: 100,
+          type: IntegrationType.Embed,
+        });
+        setEmbedIntegrations(integrations.orderedData);
```
Comment (tommoor) [inline app/hooks/useEmbedIntegrations.ts:20]: I don't think there's any need for the local state in here

File: app/scenes/Settings/Drawio.tsx [hunk 28]
```diff
@@ -43,44 +47,43 @@ function Drawio() {
...
+          id: integration?.id,
           type: IntegrationType.Embed,
-          service: "diagrams",
+          service: SERVICE_NAME,
           settings: {
             url: data.url,
           },
         });
+
+        showToast("Settings saved", {
```
Comment (tommoor) [inline app/scenes/Settings/Drawio.tsx:60]:
```suggestion
        showToast(t("Settings saved"), {
```

File: app/scenes/Settings/Drawio.tsx [hunk 29]
```diff
@@ -43,44 +47,43 @@ function Drawio() {
...
+          type: "success",
+        });
+      } catch (err) {
+        showToast("Failed to save!", {
```
Comment (tommoor) [inline app/scenes/Settings/Drawio.tsx:64]:
```suggestion
        showToast(err.message, {
```

File: app/scenes/Settings/Drawio.tsx [hunk 30]
```diff
@@ -43,44 +47,43 @@ function Drawio() {
...
+          type: "error",
+        });
       }
     },
-    [integrations, integration]
+    [integrations, integration, showToast]
```
Comment (tommoor) [inline app/scenes/Settings/Drawio.tsx:69]:
```suggestion
    [integrations, integration, t, showToast]
```

File: app/hooks/useEmbedIntegrations.ts (hunk 27 above)
Comment (apoorv-mishra) [reply app/hooks/useEmbedIntegrations.ts]: Fixed!

File: app/scenes/Settings/Drawio.tsx (hunk 28 above)
Comment (apoorv-mishra) [reply app/scenes/Settings/Drawio.tsx]: Fixed!

File: app/scenes/Settings/Drawio.tsx (hunk 29 above)
Comment (apoorv-mishra) [reply app/scenes/Settings/Drawio.tsx]: Fixed!

File: app/scenes/Settings/Drawio.tsx (hunk 30 above)
Comment (apoorv-mishra) [reply app/scenes/Settings/Drawio.tsx]: Fixed!

**Review by tommoor — APPROVED** [review]

**Comment by helmut72** [comment]
> Enables admins to allow adding hosts for self hosted embeds like diagrams.net, **grist** etc...

Is this for draw.io only or also for grist?

**Comment by tommoor** [comment]
draw.io currently, the same framework can be used for other integrations in the future though
